plugins/module_utils/inventory.py

```diff
@@ -0,0 +1,176 @@
+            if credential[1]:
+                boto_params[credential[0]] = credential[1]
+
+        return boto_params
+
+    def _set_credentials(self, templar):
+        '''
+            :param config_data: contents of the inventory config file
+        '''
+
+        credentials = {}
+
+        for credential_type in ('aws_profile', 'aws_access_key', 'aws_secret_key', 'aws_security_token', 'iam_role_arn'):
+            if templar.is_template(self.get_option(credential_type)):
+                credentials[credential_type] = templar.template(variable=self.get_option(credential_type), disable_lookups=False)
+            else:
+                credentials[credential_type] = self.get_option(credential_type)
+
+        self.boto_profile = credentials['aws_profile']
+        self.aws_access_key_id = credentials['aws_access_key']
+        self.aws_secret_access_key = credentials['aws_secret_key']
+        self.aws_security_token = credentials['aws_security_token']
+        self.iam_role_arn = credentials['iam_role_arn']
+
+        if not self.boto_profile and not (self.aws_access_key_id and self.aws_secret_access_key):
+            session = botocore.session.get_session()
+            try:
+                credentials = session.get_credentials().get_frozen_credentials()
+            except AttributeError:
+                pass
+            else:
+                self.aws_access_key_id = credentials.access_key
+                self.aws_secret_access_key = credentials.secret_key
+                self.aws_security_token = credentials.token
+
+        if not self.boto_profile and not (self.aws_access_key_id and self.aws_secret_access_key):
+            raise AWSInventoryFailure("Insufficient boto credentials found. Please provide them in your "
+                                      "inventory configuration file or set them as environment variables.")
+
+    def _boto3_assume_role(self, credentials, service, region=None):
+        """
+        Assume an IAM role passed by iam_role_arn parameter
+
+        :return: a dict containing the credentials of the assumed role
+        """
+
+        iam_role_arn = self.iam_role_arn
+
+        try:
+            sts_connection = boto3.session.Session(profile_name=self.boto_profile).client('sts', region, **credentials)
+            role_session_name = f"ansible_aws_{service}_dynamic_inventory"
+            sts_session = sts_connection.assume_role(RoleArn=iam_role_arn, RoleSessionName=role_session_name)
+            return dict(
+                aws_access_key_id=sts_session['Credentials']['AccessKeyId'],
+                aws_secret_access_key=sts_session['Credentials']['SecretAccessKey'],
+                aws_session_token=sts_session['Credentials']['SessionToken']
+            )
+        except botocore.exceptions.ClientError as e:
+            raise AWSInventoryFailure("Unable to assume IAM role: %s" % to_native(e))
+
+    def _get_connection(self, credentials, service, region='us-east-1'):
```

Comment:
A large chunk of this code feels like a rewrite of boto3_conn

I've been attacking the lookup modules over in #1225, ``plugin_utils.botocore.boto3_conn`` might be reusable...